### PR TITLE
Percent expression

### PR DESCRIPTION
**Pull request by dmingazov**

### CODE REVIEW

File: src/math/calculate/number-type.js [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    power(other) {
+        const { type } = other;
+        return type === NUMBER_TYPE && new NumberType(this.value ** other.value);
+    }
+
+    negate() {
+        return new NumberType(-this.value);
+    }
+    toString() {
+        return this.value;
```
Comment (ttypic) [inline src/math/calculate/number-type.js:39]: I think it's better to return actual string in this method, not a number

File: src/math/calculate/percent-type.js [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    multiply(other) {
+        const { type } = other;
+
+        switch (type) {
+            case NUMBER_TYPE:
+                return new NumberType(precise((this.value / 100) * other.value));
+            case PERCENT_TYPE:
+                return new NumberType(precise((this.value / 100) * (other.value / 100)));
+            default:
+                throw new Error(`Can not handle specified value type ${type}`);
```
[the remark this answers is not in the document]
Comment (dmingazov) [reply src/math/calculate/percent-type.js:44]: What is the reason for that?
not here nor in the calculate we have any clue how to handle this error, so the right decision is propagating it further

File: src/math/calculate/operators.js [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    {
+        symbol: '/',
+        name: 'divide'
+    },
+    {
+        symbol: '^',
+        name: 'power'
+    },
+    {
+        symbol: 'of',
```
[the remark this answers is not in the document]
Comment (dmingazov) [reply src/math/calculate/operators.js:26]: Still, in the context of this app, they are operators. I'll move them to a separate file though

File: src/math/calculate/operators.js [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+        name: 'partOf'
+    },
+    {
+        symbol: 'is',
+        name: 'whatNumberOf'
```
[the remark this answers is not in the document]
Comment (dmingazov) [reply src/math/calculate/operators.js]: renamed those

File: src/math/calculate/number-type.js [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    }
+}
+
+export const createNumber = val => new NumberType(parseNumber(val));
```
[the remark this answers is not in the document]
Comment (dmingazov) [reply src/math/calculate/number-type.js:39]: Any ideas where?
I would leave it here for now at least

File: src/math/interpreter.js [hunk 6]
```diff
@@ -79,22 +81,32 @@ class MathInterpreter extends BaseCstVisitor {
...
+            return calculate(val, acc, symbolToOperator['^']);
+        }, createNumber(1));
+
+        return calculate(base, reduced, symbolToOperator['^']);
+    }
+
+    whatNumberOfPercentExpression(ctx) {
+        const percent = createPercent(ctx.PercentLiteral[0].image);
+        const number = createNumber(ctx.NumberLiteral[0].image);
+        return calculate(number, percent, symbolToOperator['is']);
```
Comment (ttypic) [inline src/math/interpreter.js:103]: I think that just `percent.r_isof(number)` - more clear.

**Pull request merged**